### PR TITLE
[16.0][ADD] pos_partner_location_map

### PR DESCRIPTION
**Pull request by geomer198**
This module allows to select partner address directly on map. For this Google Maps are used.

### CODE REVIEW

**Review by legalsylvain — COMMENTED** [review]
I think the name of the module should contain "google".

File: pos_partner_location_map/static/src/js/googleMapInit.esm.js
Comment (em230418) [inline pos_partner_location_map/static/src/js/googleMapInit.esm.js]: You need to add check here if resonse is not falsy.

File: pos_partner_location_map/readme/DESCRIPTION.rst
Comment (em230418) [inline pos_partner_location_map/readme/DESCRIPTION.rst]: Since this module heavily depends on google, it should be renamed to something like `pos_partner_location_map_google`. Because there are other services like google maps. For example yandex maps. https://yandex.ru/dev/maps/

**Review by em230418 — CHANGES_REQUESTED** [review]

File: pos_partner_location_map/readme/DESCRIPTION.rst
Comment (isserver1) [reply pos_partner_location_map/readme/DESCRIPTION.rst:2]: > Since this module heavily depends on google, it should be renamed to something like `pos_partner_location_map_google`. Because there are other services like google maps. For example yandex maps. https://yandex.ru/dev/maps/

Hi @em230418 thank you for the feedback! This module implements core behaviour for both Google an OpenStreetMap. This is why we have opted for such a name keeping OpenStreetMap as a "Roadmap/Todo" item

Comment (ivs-cetmix) [reply pos_partner_location_map/readme/DESCRIPTION.rst]: > Since this module heavily depends on google, it should be renamed to something like `pos_partner_location_map_google`. Because there are other services like google maps. For example yandex maps. https://yandex.ru/dev/maps/
> I think the name of the module should contain "google".

Hi @em230418 @legalsylvain  thank you for the feedback! This module implements core behaviour for any map provide. Google Is just the first one on the list. This is why we have opted for such a name keeping OpenStreetMap as a "Roadmap/Todo" item  So anyone could modify this module later without a need of creating a separate one for each map provider.

File: pos_partner_location_map/__manifest__.py
Comment (ivs-cetmix) [inline pos_partner_location_map/__manifest__.py]: Show and set partner location directly on map in POS

File: pos_partner_location_map/__manifest__.py
Comment (ivs-cetmix) [inline pos_partner_location_map/__manifest__.py]: POS Partner Location Map

File: pos_partner_location_map/models/res_partner.py
Comment (ivs-cetmix) [inline pos_partner_location_map/models/res_partner.py]: This field might be misleading for a user. Because there are core Odoo fields used for the same goal.
We should parse this string into Odoo address fields instead.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: pos_partner_location_map/models/res_partner.py
Comment (geomer198) [reply pos_partner_location_map/models/res_partner.py]: Resolved

File: pos_partner_location_map/__manifest__.py
Comment (geomer198) [reply pos_partner_location_map/__manifest__.py]: Resolved

File: pos_partner_location_map/__manifest__.py
Comment (geomer198) [reply pos_partner_location_map/__manifest__.py]: Resolved

File: pos_partner_location_map/static/src/js/googleMapInit.esm.js
Comment (geomer198) [reply pos_partner_location_map/static/src/js/googleMapInit.esm.js]: Resolved

File: pos_partner_location_map/readme/DESCRIPTION.rst
Comment (em230418) [reply pos_partner_location_map/readme/DESCRIPTION.rst]: > This module implements core behaviour for any map provide. Google Is just the first one on the list.

Do you plan to maintain behaviour of every map provider that is included to this module by others (1) or just google and OSM only (2)? If (2), you need to create seperate modules.

**Review by em230418 — COMMENTED** [review]
Also noting, that there is OCA module for google map integration. https://github.com/OCA/geospatial/tree/14.0/web_view_google_map

**Comment by ivs-cetmix** [comment]
> Also noting, that there is OCA module for google map integration. https://github.com/OCA/geospatial/tree/14.0/web_view_google_map

Yup we are aware of. However this one is based on the Odoo core module base_geolocalize for better compatibility

File: pos_partner_location_map/readme/DESCRIPTION.rst
Comment (ivs-cetmix) [reply pos_partner_location_map/readme/DESCRIPTION.rst]: > > This module implements core behaviour for any map provide. Google Is just the first one on the list.
> 
> Do you plan to maintain behaviour of every map provider that is included to this module by others (1) or just google and OSM only (2)? If (2), you need to create seperate modules.

Yes, this module is designed to be extendable with other map providers as well

Comment (em230418) [reply pos_partner_location_map/readme/DESCRIPTION.rst]: Sorry, my question was incorrect.

> So anyone could modify this module later without a need of creating a separate one for each map provider.

Do you plan to maintain behaviour of every map provider that WILL BE included to this module (pos_partner_location_map) by others (1) or just google and OSM only (2)? If (2), you need to create separate modules.

> Yes, this module is designed to be extendable with other map providers as well

I don't have any doubt on this.

Comment (ivs-cetmix) [reply pos_partner_location_map/readme/DESCRIPTION.rst]: > Sorry, my question was incorrect.
> 
> > So anyone could modify this module later without a need of creating a separate one for each map provider.
> 
> Do you plan to maintain behaviour of every map provider that WILL BE included to this module (pos_partner_location_map) by others (1) or just google and OSM only (2)? If (2), you need to create separate modules.
> 
> > Yes, this module is designed to be extendable with other map providers as well
> 
> I don't have any doubt on this.

Yes if someone will open a PR with new map providers we will merge it when reviewed and approved.

**Comment by geomer198** [comment]
@isserver1, @legalsylvain, @em230418 Could you make code review please?

**Comment by legalsylvain** [comment]
> Hi @em230418 @legalsylvain thank you for the feedback! This module implements core behaviour for any map provide. Google Is just the first one on the list. This is why we have opted for such a name keeping OpenStreetMap as a "Roadmap/Todo" item So anyone could modify this module later without a need of creating a separate one for each map provider.

Not totally agree with that point. I think that if openstreetmap is in the roadmap, many modules should exists : 
- pos_partner_location_abstract (for common things)
- pos_partner_location_google_map (for specific things).
- pos_partner_location_openstreetmap (for specific things).

The module you provide is quite complex. (that is normal). So if tomorrow openstreetmap is implemented in the same module, we'll have a lot of useless code installed, that is not a good idea. (javascript asset related to google, or file like this address_google_struct.py). For maintenability reason also. If we implement osm in the same module as google, and a contributor want to migrate the module in version 17, the current design will ask to the contributor to migrate both features, even if one of the feature is not desired by the contributors.

At this step, I see two possibilities : 
A) creating 2 modules, but maybe it's quite hard to split and we maybe don't know what is "common" at this step.
B) rename the current module into ``pos_partner_location_google_map`` and if an other contributor want to create a openstreetmap module, it can contribute ``pos_partner_location_openstreetmap_map`` and create a new module abstract  and make the two modules (google and osm) depends on the new abstract module.

What do you think ?

**Comment by geomer198** [comment]
> > Hi @em230418 @legalsylvain thank you for the feedback! This module implements core behaviour for any map provide. Google Is just the first one on the list. This is why we have opted for such a name keeping OpenStreetMap as a "Roadmap/Todo" item So anyone could modify this module later without a need of creating a separate one for each map provider.
> 
> Not totally agree with that point. I think that if openstreetmap is in the roadmap, many modules should exists :
> 
>     * pos_partner_location_abstract (for common things)
> 
>     * pos_partner_location_google_map (for specific things).
> 
>     * pos_partner_location_openstreetmap (for specific things).
> 
> 
> The module you provide is quite complex. (that is normal). So if tomorrow openstreetmap is implemented in the same module, we'll have a lot of useless code installed, that is not a good idea. (javascript asset related to google, or file like this address_google_struct.py). For maintenability reason also. If we implement osm in the same module as google, and a contributor want to migrate the module in version 17, the current design will ask to the contributor to migrate both features, even if one of the feature is not desired by the contributors.
> 
> At this step, I see two possibilities : A) creating 2 modules, but maybe it's quite hard to split and we maybe don't know what is "common" at this step. B) rename the current module into `pos_partner_location_google_map` and if an other contributor want to create a openstreetmap module, it can contribute `pos_partner_location_openstreetmap_map` and create a new module abstract and make the two modules (google and osm) depends on the new abstract module.
> 
> What do you think ?

I think that we can split the module into two separate ones  `pos_partner_location_base` and `pos_partner_location_google_map`.
Can we keep both In the same pr or should make separate for each of them?

**Comment by legalsylvain** [comment]
>  Can we keep both In the same pr or should make separate for each of them?

In general, in the OCA, it's recommended to separate PRs. however, in this case, module A doesn't make sense without module B, and vice versa. so i don't see any advantage in separating, you can do the refactor in the same branch. That will be easier.

Thanks a lot!

**Review by GabbasovDinar — APPROVED** [review]
Not testes. code LGTM, please squash commit by modules

**Review by ivs-cetmix — APPROVED** [review]
Functional ok

**Comment by ivs-cetmix** [comment]
hey @legalsylvain @em230418 could you please check again? Would be great to have it merged

**Comment by ivs-cetmix** [comment]
@legalsylvain @em230418 I'm really sorry for bothering just didn't want to get it stalled

**Comment by legalsylvain** [comment]
Sorry for the delay. I can take a look on runboat, but there is no link. Could you rebase, to relaunch the CI ? 
thanks !

File: pos_partner_location_abstract/models/res_config_settings.py
Comment (legalsylvain) [inline pos_partner_location_abstract/models/res_config_settings.py]: Question : 
This fields looks to be not used (and not available in any view).
Did I missed something ?

File: pos_partner_location_abstract/models/pos_config.py
Comment (legalsylvain) [inline pos_partner_location_abstract/models/pos_config.py:10]: If I understand this field correctly, this should be a Selection field, don't you think ? (openstreetmap / googlemap). don't you think ?

File: pos_partner_location_abstract/models/pos_config.py
Comment (legalsylvain) [inline pos_partner_location_abstract/models/pos_config.py:19]: I don't understand very well why both functions are required.

File: pos_partner_location_abstract/models/pos_session.py
Comment (legalsylvain) [inline pos_partner_location_abstract/models/pos_session.py]: nitpicking.

```suggestion
        res = super()._loader_params_res_partner()
```

**Review by legalsylvain — COMMENTED** [review]
General review : No functional review finally, I have no Google API key. So feature doesn't work on runboat.

Nitpicking + non blocking : 
The lat / long are not well displayed in small screen. other fields added after will be maybe not well displayed. Maybe better to follow the classic rules "one field per line".

image

Nothing to do with your PR : 
on runboat, I can not change data on partner form view (in PoS) and save. (changes are discarded).
Don't know why, but the bug is present also in runboat of 16.0 branch. so unrelated with your contribution. (but annoying !)

File: pos_partner_location_google_map/static/src/js/PartnerDetailsMapGoogleEdit.esm.js
Comment (em230418) [inline pos_partner_location_google_map/static/src/js/PartnerDetailsMapGoogleEdit.esm.js:16]: Function call must be returned, not function itself.

File: pos_partner_location_abstract/static/src/js/PartnerMapEdit.esm.js
Comment (em230418) [inline pos_partner_location_abstract/static/src/js/PartnerMapEdit.esm.js]: Returned value is not used. No need to return anything.

File: pos_partner_location_google_map/static/src/js/PartnerMapGoogleEdit.esm.js
Comment (em230418) [inline pos_partner_location_google_map/static/src/js/PartnerMapGoogleEdit.esm.js]: 1. geocode returns formatted address in results[0].formatted_address and here we also do something to get formatted address using "getFormattedAddress".
2. this.address inside body getFormattedAddress could be set in 5 seconds if bad internet connection, meanwhile this.addrInput.el.value is set very quickly. Is it fine here? You are sure, we don't need to "await" for getFormattedAddress?

File: pos_partner_location_google_map/static/src/js/PartnerMapGoogleEdit.esm.js
Comment (em230418) [inline pos_partner_location_google_map/static/src/js/PartnerMapGoogleEdit.esm.js]: Please use "camelCase", not "snake_case". Since in js snake_case naming for methods is not popular.

File: pos_partner_location_google_map/models/base_geocoder.py
Comment (em230418) [inline pos_partner_location_google_map/models/base_geocoder.py]: On bad status, there should be warning indicated. At least in logs.

**Review by em230418 — CHANGES_REQUESTED** [review]

File: pos_partner_location_abstract/models/pos_config.py
Comment (geomer198) [reply pos_partner_location_abstract/models/pos_config.py:10]: > If I understand this field correctly, this should be a Selection field, don't you think ? (openstreetmap / googlemap). don't you think ?

I thinks it's not better idea. This field is compute by `Geo Provider` configuration parameter. Provider we can update or add new what raise an error, that is why this field is just Char.

File: pos_partner_location_abstract/models/pos_config.py
Comment (geomer198) [reply pos_partner_location_abstract/models/pos_config.py:19]: > I don't understand very well why both functions are required.

These functions helping fast and without problems configure new providers for POS.

File: pos_partner_location_google_map/static/src/js/PartnerDetailsMapGoogleEdit.esm.js
Comment (geomer198) [reply pos_partner_location_google_map/static/src/js/PartnerDetailsMapGoogleEdit.esm.js:16]: > Function call must be returned, not function itself.

This function is a property, that is why `super.accessToMap`  returned property value.

File: pos_partner_location_google_map/static/src/js/PartnerMapGoogleEdit.esm.js
Comment (geomer198) [reply pos_partner_location_google_map/static/src/js/PartnerMapGoogleEdit.esm.js]: Thanks, I will change it.

**Comment by geomer198** [comment]
@legalsylvain @em230418 Could you check changes?

File: pos_partner_location_google_map/static/src/js/PartnerDetailsMapGoogleEdit.esm.js
Comment (em230418) [reply pos_partner_location_google_map/static/src/js/PartnerDetailsMapGoogleEdit.esm.js:16]: I see.

File: pos_partner_location_abstract/readme/DESCRIPTION.rst
Comment (ivs-cetmix) [inline pos_partner_location_abstract/readme/DESCRIPTION.rst]: This module implements base functionality for using different map providers in POS.
It allows to select partner location directly on a map and stores selected location in the res.partner model.
It also tries to evaluate partner's address from coordinates and provides a helper function that generates location QR code.
NB: This is a core module which is meant to be used in further customisations.

File: pos_partner_location_abstract/models/res_partner.py
Comment (ivs-cetmix) [inline pos_partner_location_abstract/models/res_partner.py]: This is a non stored field. So using "write" will do no benefit here because no db tables are updated. 
You should use "update" instead.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]

File: pos_partner_location_abstract/readme/DESCRIPTION.rst
Comment (geomer198) [reply pos_partner_location_abstract/readme/DESCRIPTION.rst]: Description is updated

File: pos_partner_location_abstract/models/res_partner.py
Comment (geomer198) [reply pos_partner_location_abstract/models/res_partner.py]: Agree with you, applied proposed changes.

**Review by DemchukM — COMMENTED** [review]
Please add author and license information to all files.
Everything else looks good

**Review by ivs-cetmix — APPROVED** [review]
Finally LGTM

**Comment by legalsylvain** [comment]
@em230418, could you update your review ?

**Review by DemchukM — APPROVED** [review]
LGTM

**Comment by ivs-cetmix** [comment]
Hey @em230418 happy new year! Waiting for your updated review 😄

File: pos_partner_location_abstract/static/src/js/PartnerDetailsMapEdit.esm.js
Comment (em230418) [inline pos_partner_location_abstract/static/src/js/PartnerDetailsMapEdit.esm.js]: Since we "await" from partnerMap, I suggest also to await "this.showPopup".

File: pos_partner_location_abstract/tests/test_pos_session.py
Comment (em230418) [inline pos_partner_location_abstract/tests/test_pos_session.py]: Do we need to override setUp here?

File: pos_partner_location_abstract/tests/test_pos_ui.py
Comment (em230418) [inline pos_partner_location_abstract/tests/test_pos_ui.py]: Why running "search" instead of validating fields?

**Review by em230418 — CHANGES_REQUESTED** [review]

**Comment by geomer198** [comment]
@em230418 Could you check changes?

**Review by em230418 — APPROVED** [review]

**Comment by ivs-cetmix** [comment]
Hi @legalsylvain , would appreciate you having a look at this when you have time. Thank you in advance!)

**Comment by legalsylvain** [comment]
> Hi @legalsylvain , would appreciate you having a look at this when you have time. Thank you in advance!)

I can not test on runboat. But there are many approvals, so : 

/ocabot merge nobump

**Pull request merged**
